Update proposal for SIP-20 to enable two-way communication between Snap and WebSockets (#149)
MetaMask/SIPs: SIPS/sip-20.md

---
sip: 20

> title: Advanced external communication and events

Author: Feels like the name is not reflecting what this SIP really is anymore? Looking for advice to change it if possible. Updated with one of the ideas I had.

status: Draft
author: David Drazic (@david0xd)
created: 2023-12-15
---

## Abstract

> This SIP proposes a new permission that enables Snaps to receive data from external sources and send data back to them.

Reviewer: Not all external sources necessarily support two-way communication.

In the initial version, this SIP proposes support for WebSockets, but it can be extended to support other data sources like blockchain events in the future.
The Snap can specify the external data source in the Snap manifest. The client then connects to the external data source and sends the data to the Snap.

> The Snap can then do whatever it wants with the data.
> The Snap can use RPC method to send data back to the external connection (e.g. websocket connection).

Reviewer: ```suggestion
The Snap can then do whatever it wants with the data, and can send data back to the external source if supported (e.g., a WebSocket connection).
```

## Motivation

> Snaps are currently limited in their ability to both receive and send data from and to external sources; they have to rely on user actions or cron jobs to fetch data, so they can't react to events in real time. Snaps also cannot use WebSocket connections for bidirectional communication, and are limited to HTTP requests.

Reviewer: ```suggestion
Snaps are currently limited in their ability to communicate with external sources; they have to rely on user actions or cron jobs to fetch data, so they can't react to events in real time. Snaps also cannot use WebSocket connections for bidirectional communication, and are limited to HTTP requests.
```

The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.

### RPC Methods

> This SIP introduces new RPC method for sending data to the external connections.

Reviewer: ```suggestion
This SIP introduces a new RPC method for sending data to the external connections (if supported).
```

#### snap_sendData

Snap can use `snap_sendData` RPC method to send data specified within request params.

> The proposed RPC method `snap_sendData` SHOULD be a restricted RPC method requiring user consent before usage via the permission system. The RPC method SHOULD only be available to Snaps.

Reviewer: I think we should just use `endowment:external-data` for this RPC method as well.

> Snap MUST specify destination which is URL identifier of an external connection (e.g. websocket). Destination MUST be exact match of some of the URLs defined in manifest's permission caveats within `endowment:external-data`.

Reviewer: In theory the same URL can be used twice, in which case it would be ambiguous which one to use.

The `snap_sendData` JSON-RPC method takes an object as parameters, which has the following properties:
- `data` (required, `string` or `binary`): A data, which MUST NOT be empty.
- `destination` (required, `string`): A data destination which is one of the WebSocket URLs specified in the manifest, which MUST NOT be empty.

Example:
```json
{
  "method": "snap_sendData",
  "params": {
    "data": {
      "foo": "bar"
    },
    "destination": "wss://example.com/text/endpoint"
  }
}
```

> The method returns a `boolean`, `true` in case of successful delivery of a message to the WebSocket, `false` if delivery was unsuccessful.

Reviewer: Should it throw an error instead if delivery failed?

 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 */
type WebSocketIncomingBinaryMessage = {
  type: 'binary';
  message: ArrayBuffer;
};

/**
 * A text message sent to a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string.
 * @property destination - The destination web socket URL as a string.
 */
type WebSocketOutgoingTextMessage = {
    type: 'text';
    message: string;
    destination: string;
};

/**
 * A binary message sent to a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 * @property destination - The destination web socket URL as a string.
 */
type WebSocketOutgoingBinaryMessage = {
    type: 'binary';
    message: ArrayBuffer;
    destination: string;

> };

Reviewer: This can be simplified:

```ts
type WebSocketIncomingMessage = WebSocketIncomingTextMessage | WebSocketIncomingBinaryMessage;
type WebSocketOutgoingMessage = WebSocketIncomingMessage & {
  destination: string;
};
```

 */
type WebSocketData = {
  type: 'websocket';

>   message: WebSocketIncomingTextMessage | WebSocketIncomingBinaryMessage;

Reviewer: I think `data` makes more sense here.
